### PR TITLE
Setting up team's Mac mini

### LINKED ISSUE #1206: `+devtools` fails to concretize on macos
```
[exe: /Users/chapman39/dev/serac/serac_libs1/spack/bin/spack -D /Users/chapman39/dev/serac/serac_libs1/spack_env concretize --fresh ]
==> Error: concretization failed for the following reasons:

   1. Cannot select a single "version" for package "py-numpy"
[spack version: 0.22.0.dev0 (6d244b3f6776b3697861c6b5cebbee376c30bff3)
]
```

Only when I remove the python section from my spack.yaml, I'm able to concretize with `+devtools`.

_Originally posted by @chapman39 in https://github.com/LLNL/serac/issues/1187#issuecomment-2259026925_

### LINKED ISSUE #1207: `boundary_cond` test fails on macos
Looks like boundary condition `boundary_cond` test fails. I can look into this a bit more but wanted to report.

```
***********************************
[WARNING in line 67 of file /Users/chapman39/dev/serac/repo/src/serac/infrastructure/../../serac/physics/boundary_conditions/boundary_condition.hpp]
MESSAGE=Attempting to compare tags of two different enum types (always false)
***********************************

...

[  FAILED  ] 1 test, listed below:
[  FAILED  ] BoundaryCond.DirectTrueDofs
```

_Originally posted by @chapman39 in https://github.com/LLNL/serac/issues/1187#issuecomment-2259403916_

### PR DESCRIPTION
**Pull request by chapman39**
Some things I've updated (am updating) during the process of setting up the Serac shared Mac mini.

Features:

- [x] Update example mac `spack.yaml`
  - [x] Separate Brew packages from `/usr` packages
- [x] Update docs on how to build on Mac (package versions, adding additional packages for simpler concretizer tree, etc.)
- [x] Fixes Mac devtool issues  by creating a Python venv before calling uberenv (Fixes #1206)
- [x] Removes BoundaryCond.DirectTrueDofs test, which was failing on MacMini only (Fixes #1207)
- [x] Setup cron job testing Serac build/ tests
- [x] Document how to setup an additional user on the Mac mini

Readthedocs: 
- https://serac.readthedocs.io/en/task-chapman39-macmini-build/sphinx/quickstart.html#building-serac-dependencies-on-macos
- https://serac.readthedocs.io/en/task-chapman39-macmini-build/sphinx/dev_guide/macmini.html

### CODE REVIEW

**Comment by chapman39** [comment]
Installing the latest llvm/ clang (@19.1.1) caused some problems building Axom, so to avoid that used `brew install llvm@14`. Here is the build error I got:

```
==> Installing axom-0.9.0.1-c2vvp4ki46aesbfiafjfnhpil35x3hqr [37/39]
==> No binary for axom-0.9.0.1-c2vvp4ki46aesbfiafjfnhpil35x3hqr found: installing from source
==> Ran patch() for axom
==> axom: Executing phase: 'initconfig'
==> axom: Executing phase: 'cmake'
==> axom: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j10'

...

  >> 370    /Users/chapman39/dev/serac/serac_tpls/spack/opt/spack/darwin-sonoma-m2/clang-19.1.1/raja-2024.02.0-cvabateje3nivxumnufmwkvhqdhawjww/include/RAJA/pattern/kernel/Tile.hpp:174:30: error: no member named 'block_id' in 'I
            terableTiler<Iterable>'
     371      174 |       return block_id != rhs.block_id;
     372          |                          ~~~ ^
```

**Comment by chapman39** [comment]
I think I resolved my python issues. Now I'm getting this...

```
firion:build-firion-darwin-sonoma-aarch64-clang@14.0.6-debug chapman39$ make -j
ld: warning: reexported library with install name '/opt/homebrew/opt/llvm@14/lib/libunwind.1.dylib' found at '/opt/homebrew/Cellar/llvm@14/14.0.6/lib/libunwind.1.0.dylib' couldn't be matched with any parent library and will be linked directly
...
[ 11%] Linking CXX executable ../../tests/axom_smoketest
ld: warning: duplicate -rpath '/Users/chapman39/dev/serac/serac_tpls/spack/opt/spack/darwin-sonoma-aarch64/clang-14.0.6/hypre-2.26.0-4pntdjmzmzrwgopttqkxwueelpjy7kzp/lib' ignored
ld: warning: duplicate -rpath '/Users/chapman39/dev/serac/serac_tpls/spack/opt/spack/darwin-sonoma-aarch64/clang-14.0.6/superlu-dist-8.1.2-6elfwrr5vzdogxmamzvy5pcdpuzovj6t/lib' ignored
...
ld: warning: could not create compact unwind for ___slepcsys_MOD_slepcinitializenoarguments: does not use standard frame
...
ld: warning: could not create compact unwind for _petscoptionsenum_: does not use standard frame
ld: warning: could not create compact unwind for _f90arraysfnodecreate_: does not use standard frame
```

**Comment by chapman39** [comment]
Current issue is when the macos script is called via cron I get `dot` not found errors. However when I call it via command line myself it works just fine. And `which dot` works.

**Comment by chapman39** [comment]
I should also ensure an email is sent even if the build script fails.

**Review by tupek2 — APPROVED** [review]

**Comment by chapman39** [comment]
`functional_basic_dg` has failed in CI for me twice today. Retry resolves it but just wanted to let you know @samuelpmishLLNL

**Comment by chapman39** [comment]
@btalamini Here is a summary of some of the linker warnings the mac mini build gets

```
 9013 ld: warning: reexported library with install name '/opt/homebrew/opt/llvm@14/lib/libunwind.1.dylib' found at '/opt/homebrew/Cellar/llvm@14/14.0.6/lib/libunwind.1.0.dylib' couldn't be matched with any parent library and will be linked directly
...
 9143 ld: warning: -ld_classic is deprecated and will be removed in a future release
...
 9046 ld: warning: duplicate -rpath '/opt/homebrew/lib' ignored
 9047 ld: warning: duplicate -rpath '/Users/chapman39/dev/serac/serac_tpls/spack/opt/spack/darwin-sonoma-aarch64/clang-14.0.6/parmetis-4.0.3-faxw36ybmuntuwxu5p3zka57sbf3yyfe/lib' ignored
...
 9079 ld: warning: could not create compact unwind for ___gcc_nested_func_ptr_deleted: does not use standard frame
 9080 ld: warning: could not create compact unwind for _petscbagregisterenum_: does not use standard frame
 9081 ld: warning: could not create compact unwind for ___petscsys_MOD_petscinitializenoarguments: does not use standard frame
```

File: scripts/shared-macmini/build-and-test.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+PROJECT_DIR="$CI_ROOT_DIR/repo"
+OUTPUT_LOG="$CI_ROOT_DIR/logs/macmini-build-and-test-$(date +"%Y_%m_%d_%H_%M_%S").log"
+HOST_CONFIG="$CI_ROOT_DIR/host-configs/firion-darwin-sonoma-aarch64-clang@14.0.6.cmake"
+RECIPIENTS="chapman39@llnl.gov,white238@llnl.gov,talamini1@llnl.gov"
+
+# Go to project directory
+cd $PROJECT_DIR
+
+# Update Serac
+git checkout task/chapman39/macmini-build >> $OUTPUT_LOG 2>&1 # TODO CHANGE TO DEVELOP
```
Comment (chapman39) [inline scripts/shared-macmini/build-and-test.sh]: note to self to change this before merging

File: src/docs/sphinx/dev_guide/macmini.rst [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+Contact LivIT directly to setup an appointment and request one.
+
+3. **EARS Admin Request**
+
+Next, request admin access to the machine by visiting either ServiceNow or the `EARS website <https://ears.llnl.gov/dashboard>`_.
+
+4. **Two Logins**
+
+Once you have a MyPass and you have admin rights, try to log in again. There should be two passwords required to log in. The first one
+is for your MyPass (assuming it's connected to the machine) and the other is for the account login.
```
Comment (chapman39) [inline src/docs/sphinx/dev_guide/macmini.rst]:
```suggestion
is for your MyPass (assuming it's plugged into to the machine) and the other is for the account login.
```

File: scripts/shared-macmini/build-and-test.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+git pull >> $OUTPUT_LOG 2>&1
+git submodule update --init --recursive >> $OUTPUT_LOG 2>&1
+
+# Clear previous build(s)
+rm -rfv _serac_build_and_test* >> $OUTPUT_LOG 2>&1
```
Comment (chapman39) [inline scripts/shared-macmini/build-and-test.sh]:
```suggestion
rm -rf _serac_build_and_test* >> $OUTPUT_LOG 2>&1
```

**Comment by tupek2** [comment]
> Installing the latest llvm/ clang (@19.1.1) caused some problems building Axom, so to avoid that used `brew install llvm@14`. Here is the build error I got:
> 
> ```
> ==> Installing axom-0.9.0.1-c2vvp4ki46aesbfiafjfnhpil35x3hqr [37/39]
> ==> No binary for axom-0.9.0.1-c2vvp4ki46aesbfiafjfnhpil35x3hqr found: installing from source
> ==> Ran patch() for axom
> ==> axom: Executing phase: 'initconfig'
> ==> axom: Executing phase: 'cmake'
> ==> axom: Executing phase: 'build'
> ==> Error: ProcessError: Command exited with status 2:
>     'make' '-j10'
> 
> ...
> 
>   >> 370    /Users/chapman39/dev/serac/serac_tpls/spack/opt/spack/darwin-sonoma-m2/clang-19.1.1/raja-2024.02.0-cvabateje3nivxumnufmwkvhqdhawjww/include/RAJA/pattern/kernel/Tile.hpp:174:30: error: no member named 'block_id' in 'I
>             terableTiler<Iterable>'
>      371      174 |       return block_id != rhs.block_id;
>      372          |                          ~~~ ^
> ```

FYI, I was able to get things compiled with llvm19.1.6 by pointing instead to ^raja@main

**Comment by chapman39** [comment]
> FYI, I was able to get things compiled with llvm19.1.6 by pointing instead to ^raja@main

@tupek2 i made an issue here https://github.com/LLNL/radiuss-spack-configs/issues/115 we should update spack and radiuss-spack-configs next tpl build

EDIT: nvm. updating to raja@2024.07.0 will resolve the issue, and that version is already in radiuss-spack-configs. we may need to update the radiuss-spack-configs submodule though.

File: scripts/shared-macmini/build-and-test.sh (hunk 1 above)
Comment (white238) [inline scripts/shared-macmini/build-and-test.sh]: It should be helpful when reading the log to have some large comment breaks. If things go wrong it helps associating errors with commands. Something like :

```
########### Git Checkout #########
// git command
########### End Git Checkout ######
```

File: src/docs/sphinx/quickstart.rst [hunk 4]
```diff
@@ -325,35 +329,70 @@ Note that the version format output by the above command is not the same as that
 If you are not using an M2 or M3 Mac, you will need to change the ``target`` for the compiler to ``x86_64`` or ``aarch64`` for Intel and M1-based Macs, respectively.
 Similarly, you need to set the ``operating_system`` to the proper value if you are not using ``sonoma`` (MacOS 14.X).
 
-If you want to install the devtools, you should also add the following under ``packages`` in the ``spack.yaml`` files.
+Installing Python Developer Tools
```
Comment (white238) [inline src/docs/sphinx/quickstart.rst]: "Python Developer Tools" to me means that we are developing Python, which we aren't. Maybe change this to just "Project Developer Tools"

File: src/docs/sphinx/quickstart.rst [hunk 5]
```diff
@@ -325,35 +329,70 @@ Note that the version format output by the above command is not the same as that
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This step is only required if you wish to use Serac's developer tools. In order to use Python devtools, you will need to create a Python venv. This is much more reliable than having Spack install 20+ Python packages.
+In this example, we are using the builtin Python in ``/usr/bin``, but it is possible to use a version installed from Brew or elsewhere.
+
+Next, you will need to install wheel, sphinx, and `ATS <https://github.com/LLNL/ATS/tree/7.0.105>`_:
```
Comment (white238) [inline src/docs/sphinx/quickstart.rst]:
```suggestion
Next, you will need to install wheel, Sphinx, and `ATS <https://github.com/LLNL/ATS/tree/7.0.105>`_:
```

File: src/docs/sphinx/quickstart.rst [hunk 6]
```diff
@@ -325,35 +329,70 @@ Note that the version format output by the above command is not the same as that
...
+.. code-block:: bash
+
+   python3 -m venv --system-site-packages venv
+   source venv/bin/activate
+   pip install wheel
+   pip install sphinx
+   git clone git@github.com:LLNL/ATS.git --branch 7.0.105
+   pip install ATS/
+
+Keep track of the sphinx version while installing, since you'll need it for the next step.
```
Comment (white238) [inline src/docs/sphinx/quickstart.rst]:
```suggestion
Keep track of the Sphinx version while installing, since you'll need it for the next step.
```

File: src/docs/sphinx/quickstart.rst [hunk 7]
```diff
@@ -325,35 +329,70 @@ Note that the version format output by the above command is not the same as that
...
+      buildable: false
+      externals:
+      - spec: py-ats@7.0.105
+        prefix: /path/to/venv
 
 Building dependencies
 ^^^^^^^^^^^^^^^^^^^^^
 
-The invocation of ``uberenv.py`` is slightly modified from the standard instructions above in order to force the use of the Homebrew-installed MPI and compilers:
+Uberenv is a Spack wrapper simplifing the TPL build process. The invocation of ``uberenv.py`` is slightly modified from the standard instructions above in order to force the use of the Homebrew-installed MPI and compilers:
```
Comment (white238) [inline src/docs/sphinx/quickstart.rst:391]: There is a mix between "Brew" and "Homebrew". Please pick one

**Review by white238 — APPROVED** [review]
Small nitpicks but otherwise looks great. Thanks @chapman39 !

File: scripts/shared-macmini/build-and-test.sh (hunk 1 above)
Comment (chapman39) [reply scripts/shared-macmini/build-and-test.sh]: ended up with a slightly different approach - wrapping printing, running, logging commands in a function. heres a sample

```
####################
# git pull
####################
X11 forwarding request failed on channel 0
Already up to date.

####################
# git submodule update --init --recursive
####################
error: Your local changes to the following files would be overwritten by checkout:
	uberenv.py
Please commit your changes or stash them before you switch branches.
Aborting
Unable to checkout '0a39ce245d7866374bf4724bec9da6ab4cf4dfcc' in submodule path 'scripts/uberenv'

####################
# mutt -a /usr/workspace/meemee/serac/repo/macmini-build-and-test-2025_01_17_13_46_52.log -s Serac Succeeded! MacMini build and test report Fri Jan 17 13:46:57 PST 2025 -- chapman39@llnl.gov
####################
```
